File: agent-go/mcp/toolkit_test.go
package mcp_test

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"net"
	"net/http"
	"os"
	"os/exec"
	"syscall"
	"testing"
	"time"

	"github.com/google/go-cmp/cmp"
	llmagent "github.com/hoangvvo/llm-sdk/agent-go"
	llmmcp "github.com/hoangvvo/llm-sdk/agent-go/mcp"
	llmsdk "github.com/hoangvvo/llm-sdk/sdk-go"
	llmsdktest "github.com/hoangvvo/llm-sdk/sdk-go/llmsdktest"
	gomcp "github.com/modelcontextprotocol/go-sdk/mcp"
)

type listArgs struct {
	Shift string `json:"shift" jsonschema_description:"Which operating window to query." jsonschema_enum:"evening,overnight"`
}

type toolResponder func(args listArgs) *gomcp.CallToolResult

type stubServer struct {
	url        string
	stop       func()
	updateTool func(name, description string, responder toolResponder)
}

var (
	imageBytes  = []byte{0x00, 0x01, 0x02}
	audioBytes  = []byte{0x03, 0x04}
	imageBase64 = base64.StdEncoding.EncodeToString(imageBytes)
	audioBase64 = base64.StdEncoding.EncodeToString(audioBytes)
)

func startStubMCPServer() (*stubServer, error) {
	server := gomcp.NewServer(&gomcp.Implementation{Name: "stub-mcp", Version: "1.0.0"}, nil)

	toolName := "list_shuttles"
	toolDescription := "List active shuttle routes for a shift"
	responder := func(args listArgs) *gomcp.CallToolResult {
		summary := "Shuttle summary for " + args.Shift + " shift."
		return &gomcp.CallToolResult{
			Content: []gomcp.Content{
				&gomcp.TextContent{Text: summary},
				&gomcp.ImageContent{MIMEType: "image/png", Data: imageBytes},
				&gomcp.AudioContent{MIMEType: "audio/mpeg", Data: audioBytes},
				&gomcp.ResourceLink{URI: "https://example.com/docs", Name: "ignored"},
			},
		}
	}

	registerTool := func() {
		gomcp.AddTool(server, &gomcp.Tool{
			Name:        toolName,
			Description: toolDescription,
		}, func(ctx context.Context, _ *gomcp.CallToolRequest, args listArgs) (*gomcp.CallToolResult, any, error) {
			_ = ctx
			return responder(args), nil, nil
		})
	}

	registerTool()

	handler := gomcp.NewStreamableHTTPHandler(
		func(*http.Request) *gomcp.Server { return server },
		&gomcp.StreamableHTTPOptions{Stateless: true, JSONResponse: true},
	)

	listener, err := net.Listen("tcp4", "127.0.0.1:0")
	if err != nil {
		return nil, err
	}

	srv := &http.Server{Handler: handler}
	go func() {
		if serveErr := srv.Serve(listener); serveErr != nil && !errors.Is(serveErr, http.ErrServerClosed) {
			// best effort logging via testing output when available
		}
	}()

	cleanup := func() {
		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		defer cancel()
		_ = srv.Shutdown(ctx)
	}

	update := func(name, description string, nextResponder toolResponder) {
		server.RemoveTools(toolName)
		if name != "" {
			toolName = name
		}
		if description != "" {
			toolDescription = description
		}
		if nextResponder != nil {
			responder = nextResponder
		}
		registerTool()
	}

	return &stubServer{
		url:        "http://" + listener.Addr().String(),
		stop:       cleanup,
		updateTool: update,
	}, nil
}

func TestMCPToolkitSessionHydratesToolsAndExecutes(t *testing.T) {
	stub, err := startStubMCPServer()
	if err != nil {
		var opErr *net.OpError
		if errors.As(err, &opErr) {
			var sysErr *os.SyscallError
			if errors.As(opErr.Err, &sysErr) {
				switch sysErr.Err {
				case syscall.EACCES, syscall.EPERM:
					t.Skipf("skipping: listening on loopback blocked (%v)", sysErr.Err)
				}
			}
		}
		t.Fatalf("start stub server: %v", err)
	}
	t.Cleanup(stub.stop)

	ctx := context.Background()

	model := llmsdktest.NewMockLanguageModel()
	toolCallArgs, err := json.Marshal(map[string]string{"shift": "evening"})
	if err != nil {
		t.Fatalf("marshal tool args: %v", err)
	}
	model.EnqueueGenerateResult(llmsdktest.NewMockGenerateResultResponse(llmsdk.ModelResponse{
		Content: []llmsdk.Part{
			{
				ToolCallPart: &llmsdk.ToolCallPart{
					ToolCallID: "call_1",
					ToolName:   "list_shuttles",
					Args:       toolCallArgs,
				},
			},
		},
	}))
	model.EnqueueGenerateResult(llmsdktest.NewMockGenerateResultResponse(llmsdk.ModelResponse{
		Content: []llmsdk.Part{{TextPart: &llmsdk.TextPart{Text: "Ready to roll."}}},
	}))

	agent := llmagent.NewAgent[struct{}](
		"mcp-test",
		model,
		llmagent.WithToolkits(
			llmmcp.NewMCPToolkit(llmmcp.StaticMCPInit[struct{}](llmmcp.NewMCPStreamableHTTPParams(stub.url, ""))),
		),
	)

	session, err := agent.CreateSession(ctx, struct{}{})
	if err != nil {
		t.Fatalf("create session: %v", err)
	}
	t.Cleanup(func() {
		if cerr := session.Close(ctx); cerr != nil {
			var exitErr *exec.ExitError
			if errors.As(cerr, &exitErr) && exitErr.ExitCode() == 13 {
				return
			}
			t.Errorf("close session: %v", cerr)
		}
	})

	resp, err := session.Run(ctx, llmagent.RunSessionRequest{
		Input: []llmagent.AgentItem{
			llmagent.NewAgentItemMessage(llmsdk.NewUserMessage(llmsdk.NewTextPart("What shuttles are running tonight?"))),
		},
	})
	if err != nil {
		t.Fatalf("run session: %v", err)
	}

	expected := &llmagent.AgentResponse{
		Content: []llmsdk.Part{{TextPart: &llmsdk.TextPart{Text: "Ready to roll."}}},
		Output: []llmagent.AgentItem{
			llmagent.NewAgentItemModelResponse(llmsdk.ModelResponse{
				Content: []llmsdk.Part{
					{
						ToolCallPart: &llmsdk.ToolCallPart{
							ToolCallID: "call_1",
							ToolName:   "list_shuttles",
							Args:       toolCallArgs,
						},
					},
				},
			}),
			llmagent.NewAgentItemTool(
				"call_1",
				"list_shuttles",
				toolCallArgs,
				[]llmsdk.Part{
					{TextPart: &llmsdk.TextPart{Text: "Shuttle summary for evening shift."}},
					{ImagePart: &llmsdk.ImagePart{MimeType: "image/png", ImageData: imageBase64}},
					{AudioPart: &llmsdk.AudioPart{AudioData: audioBase64, Format: llmsdk.AudioFormatMP3}},
				},
				false,
			),
			llmagent.NewAgentItemModelResponse(llmsdk.ModelResponse{
				Content: []llmsdk.Part{{TextPart: &llmsdk.TextPart{Text: "Ready to roll."}}},
			}),
		},
	}

	if diff := cmp.Diff(expected, resp); diff != "" {
		t.Fatalf("agent response mismatch (-want +got): %s", diff)
	}
}

func TestMCPToolkitSessionRefreshesToolsOnChange(t *testing.T) {
	stub, err := startStubMCPServer()
	if err != nil {
		var opErr *net.OpError
		if errors.As(err, &opErr) {
			var sysErr *os.SyscallError
			if errors.As(opErr.Err, &sysErr) {
				switch sysErr.Err {
				case syscall.EACCES, syscall.EPERM:
					t.Skipf("skipping: listening on loopback blocked (%v)", sysErr.Err)
				}
			}
		}
		t.Fatalf("start stub server: %v", err)
	}
	t.Cleanup(stub.stop)

	ctx := context.Background()

	model := llmsdktest.NewMockLanguageModel()
	toolCallArgs, err := json.Marshal(map[string]string{"shift": "evening"})
	if err != nil {
		t.Fatalf("marshal tool args: %v", err)
	}

	model.EnqueueGenerateResult(llmsdktest.NewMockGenerateResultResponse(llmsdk.ModelResponse{
		Content: []llmsdk.Part{
			{
				ToolCallPart: &llmsdk.ToolCallPart{
					ToolCallID: "call_1",
					ToolName:   "list_shuttles",
					Args:       toolCallArgs,
				},
			},
		},
	}))
	model.EnqueueGenerateResult(llmsdktest.NewMockGenerateResultResponse(llmsdk.ModelResponse{
		Content: []llmsdk.Part{{TextPart: &llmsdk.TextPart{Text: "Ready to roll."}}},
	}))
	model.EnqueueGenerateResult(llmsdktest.NewMockGenerateResultResponse(llmsdk.ModelResponse{
		Content: []llmsdk.Part{
			{
				ToolCallPart: &llmsdk.ToolCallPart{
					ToolCallID: "call_2",
					ToolName:   "list_shuttles_v2",
					Args:       toolCallArgs,
				},
			},
		},
	}))
	model.EnqueueGenerateResult(llmsdktest.NewMockGenerateResultResponse(llmsdk.ModelResponse{
		Content: []llmsdk.Part{{TextPart: &llmsdk.TextPart{Text: "Routes synced."}}},
	}))

	agent := llmagent.NewAgent[struct{}](
		"mcp-test",
		model,
		llmagent.WithToolkits(
			llmmcp.NewMCPToolkit(llmmcp.StaticMCPInit[struct{}](llmmcp.NewMCPStreamableHTTPParams(stub.url, ""))),
		),
	)

	session, err := agent.CreateSession(ctx, struct{}{})
	if err != nil {
		t.Fatalf("create session: %v", err)
	}
	t.Cleanup(func() {
		if cerr := session.Close(ctx); cerr != nil {
			var exitErr *exec.ExitError
			if errors.As(cerr, &exitErr) && exitErr.ExitCode() == 13 {
				return
			}
			t.Errorf("close session: %v", cerr)
		}
	})

	firstResp, err := session.Run(ctx, llmagent.RunSessionRequest{
		Input: []llmagent.AgentItem{
			llmagent.NewAgentItemMessage(llmsdk.NewUserMessage(llmsdk.NewTextPart("What shuttles are running tonight?"))),
		},
	})
	if err != nil {
		t.Fatalf("first run: %v", err)
	}

	expectedFirst := &llmagent.AgentResponse{
		Content: []llmsdk.Part{{TextPart: &llmsdk.TextPart{Text: "Ready to roll."}}},
		Output: []llmagent.AgentItem{
			llmagent.NewAgentItemModelResponse(llmsdk.ModelResponse{
				Content: []llmsdk.Part{
					{
						ToolCallPart: &llmsdk.ToolCallPart{
							ToolCallID: "call_1",
							ToolName:   "list_shuttles",
							Args:       toolCallArgs,
						},
					},
				},
			}),
			llmagent.NewAgentItemTool(
				"call_1",
				"list_shuttles",
				toolCallArgs,
				[]llmsdk.Part{
					{TextPart: &llmsdk.TextPart{Text: "Shuttle summary for evening shift."}},
					{ImagePart: &llmsdk.ImagePart{MimeType: "image/png", ImageData: imageBase64}},
					{AudioPart: &llmsdk.AudioPart{AudioData: audioBase64, Format: llmsdk.AudioFormatMP3}},
				},
				false,
			),
			llmagent.NewAgentItemModelResponse(llmsdk.ModelResponse{
				Content: []llmsdk.Part{{TextPart: &llmsdk.TextPart{Text: "Ready to roll."}}},
			}),
		},
	}
	if diff := cmp.Diff(expectedFirst, firstResp); diff != "" {
		t.Fatalf("first response mismatch (-want +got): %s", diff)
	}

	stub.updateTool(
		"list_shuttles_v2",
		"List active shuttle routes with live updates",
		func(args listArgs) *gomcp.CallToolResult {
			return &gomcp.CallToolResult{
				Content: []gomcp.Content{
					&gomcp.TextContent{Text: "Updated shuttle roster for " + args.Shift + " shift."},
				},
			}
		},
	)

	time.Sleep(20 * time.Millisecond)

	secondResp, err := session.Run(ctx, llmagent.RunSessionRequest{
		Input: []llmagent.AgentItem{
			llmagent.NewAgentItemMessage(llmsdk.NewUserMessage(llmsdk.NewTextPart("How about now?"))),
		},
	})
	if err != nil {
		t.Fatalf("second run: %v", err)
	}

	expectedSecond := &llmagent.AgentResponse{
		Content: []llmsdk.Part{{TextPart: &llmsdk.TextPart{Text: "Routes synced."}}},
		Output: []llmagent.AgentItem{
			llmagent.NewAgentItemModelResponse(llmsdk.ModelResponse{
				Content: []llmsdk.Part{
					{
						ToolCallPart: &llmsdk.ToolCallPart{
							ToolCallID: "call_2",
							ToolName:   "list_shuttles_v2",
							Args:       toolCallArgs,
						},
					},
				},
			}),
			llmagent.NewAgentItemTool(
				"call_2",
				"list_shuttles_v2",
				toolCallArgs,
				[]llmsdk.Part{{TextPart: &llmsdk.TextPart{Text: "Updated shuttle roster for evening shift."}}},
				false,
			),
			llmagent.NewAgentItemModelResponse(llmsdk.ModelResponse{
				Content: []llmsdk.Part{{TextPart: &llmsdk.TextPart{Text: "Routes synced."}}},
			}),
		},
	}
	if diff := cmp.Diff(expectedSecond, secondResp); diff != "" {
		t.Fatalf("second response mismatch (-want +got): %s", diff)
	}
}

File: agent-js/src/mcp/toolkit.test.ts
/* eslint-disable @typescript-eslint/no-floating-promises */
import { randomUUID } from "node:crypto";
import {
  createServer,
  type IncomingMessage,
  type ServerResponse,
} from "node:http";
import type { AddressInfo } from "node:net";
import test, { suite, type TestContext } from "node:test";

import { Agent } from "../agent.ts";
import { convertMCPContentToParts, type MCPContent } from "./content.ts";
import { mcpToolkit } from "./toolkit.ts";

import { MockLanguageModel } from "@hoangvvo/llm-sdk/test";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StreamableHTTPServerTransport } from "@modelcontextprotocol/sdk/server/streamableHttp.js";
import { z } from "zod";
import type { AgentResponse } from "../types.ts";

const IMAGE_DATA = Buffer.from([0x00, 0x01, 0x02]).toString("base64");
const AUDIO_DATA = Buffer.from([0x03, 0x04]).toString("base64");

type ToolResponder = (params: { shift: "evening" | "overnight" }) => {
  content: (
    | { type: "text"; text: string }
    | { type: "image"; mimeType: string; data: string }
    | { type: "audio"; mimeType: string; data: string }
    | { type: "resource_link"; uri: string; name: string }
  )[];
};

async function readJsonBody(req: IncomingMessage): Promise<unknown> {
  const chunks: Buffer[] = [];

  for await (const chunk of req) {
    chunks.push(chunk as Buffer);
  }

  if (chunks.length === 0) {
    return undefined;
  }

  const body = Buffer.concat(chunks).toString();
  return body ? JSON.parse(body) : undefined;
}

async function startStubMcpServer(): Promise<{
  url: string;
  stop: () => Promise<void>;
  updateTool: (options: {
    name?: string;
    description?: string;
    responder?: ToolResponder;
  }) => void;
}> {
  const server = new McpServer({
    name: "stub-mcp",
    version: "1.0.0",
  });

  let responder: ToolResponder = ({ shift }) => ({
    content: [
      {
        type: "text",
        text: `Shuttle summary for ${shift} shift.`,
      },
      {
        type: "image",
        mimeType: "image/png",
        data: IMAGE_DATA,
      },
      {
        type: "audio",
        mimeType: "audio/mpeg",
        data: AUDIO_DATA,
      },
      {
        type: "resource_link",
        uri: "https://example.com/docs",
        name: "ignored",
      },
    ],
  });

  const registeredTool = server.registerTool(
    "list_shuttles",
    {
      description: "List active shuttle routes for a shift",
      inputSchema: {
        shift: z
          .enum(["evening", "overnight"])
          .describe("Which operating window to query."),
      },
    },
    ({ shift }: { shift: "evening" | "overnight" }) => responder({ shift }),
  );

  const transport = new StreamableHTTPServerTransport({
    sessionIdGenerator: () => randomUUID(),
    enableJsonResponse: true,
  });

  await server.connect(transport);

  const httpServer = createServer(
    // eslint-disable-next-line @typescript-eslint/no-misused-promises
    async (req: IncomingMessage, res: ServerResponse) => {
      try {
        const parsedBody = await readJsonBody(req);
        await transport.handleRequest(req, res, parsedBody);
      } catch (error) {
        res.statusCode = 500;
        res.setHeader("Content-Type", "application/json");
        res.end(
          JSON.stringify({
            error: error instanceof Error ? error.message : String(error),
          }),
        );
      }
    },
  );

  const listenPromise = new Promise<AddressInfo>((resolve, reject) => {
    httpServer.once("error", reject);
    httpServer.listen(0, "127.0.0.1", () => {
      httpServer.off("error", reject);
      resolve(httpServer.address() as AddressInfo);
    });
  });

  const address = await listenPromise;
  const baseUrl = `http://${address.address}:${String(address.port)}`;

  return {
    url: baseUrl,
    stop: async () => {
      await transport.close();
      await server.close();
      await new Promise<void>((resolve, reject) => {
        httpServer.close((err) => {
          if (err) reject(err);
          else resolve();
        });
      });
    },
    updateTool: ({ name, description, responder: nextResponder }) => {
      if (nextResponder) {
        responder = nextResponder;
      }
      registeredTool.update({
        ...(name ? { name } : {}),
        ...(description ? { description } : {}),
      });
    },
  };
}

suite("MCP toolkit", () => {
  test("agent hydrates MCP tools over streamable HTTP", async (t: TestContext) => {
    const stub = await startStubMcpServer();

    try {
      const model = new MockLanguageModel();
      model.enqueueGenerateResult({
        response: {
          content: [
            {
              type: "tool-call",
              tool_name: "list_shuttles",
              tool_call_id: "call_1",
              args: { shift: "evening" },
            },
          ],
        },
      });
      model.enqueueGenerateResult({
        response: {
          content: [{ type: "text", text: "Ready to roll." }],
        },
      });

      const agent = new Agent({
        name: "mcp-test",
        model,
        instructions: [],
        toolkits: [
          mcpToolkit(() => ({
            type: "streamable-http",
            url: stub.url,
          })),
        ],
      });

      const session = await agent.createSession({});

      try {
        const response = await session.run({
          input: [
            {
              type: "message",
              role: "user",
              content: [
                {
                  type: "text",
                  text: "What's running tonight?",
                },
              ],
            },
          ],
        });

        const expected: AgentResponse = {
          content: [{ type: "text", text: "Ready to roll." }],
          output: [
            {
              type: "model",
              content: [
                {
                  type: "tool-call",
                  tool_name: "list_shuttles",
                  tool_call_id: "call_1",
                  args: { shift: "evening" },
                },
              ],
            },
            {
              type: "tool",
              tool_name: "list_shuttles",
              tool_call_id: "call_1",
              input: { shift: "evening" },
              output: [
                { type: "text", text: "Shuttle summary for evening shift." },
                {
                  type: "image",
                  mime_type: "image/png",
                  data: IMAGE_DATA,
                },
                {
                  type: "audio",
                  data: AUDIO_DATA,
                  format: "mp3",
                },
              ],
              is_error: false,
            },
            {
              type: "model",
              content: [{ type: "text", text: "Ready to roll." }],
            },
          ],
        };
        t.assert.deepStrictEqual(response, expected);
      } finally {
        await session.close();
      }
    } finally {
      await stub.stop();
    }
  });

  test("agent refreshes tools on MCP list change", async (t: TestContext) => {
    const stub = await startStubMcpServer();

    try {
      const model = new MockLanguageModel();
      model.enqueueGenerateResult({
        response: {
          content: [
            {
              type: "tool-call",
              tool_name: "list_shuttles",
              tool_call_id: "call_1",
              args: { shift: "evening" },
            },
          ],
        },
      });
      model.enqueueGenerateResult({
        response: {
          content: [{ type: "text", text: "Ready to roll." }],
        },
      });
      model.enqueueGenerateResult({
        response: {
          content: [
            {
              type: "tool-call",
              tool_name: "list_shuttles_v2",
              tool_call_id: "call_2",
              args: { shift: "evening" },
            },
          ],
        },
      });
      model.enqueueGenerateResult({
        response: {
          content: [{ type: "text", text: "Routes synced." }],
        },
      });

      const agent = new Agent({
        name: "mcp-test",
        model,
        instructions: [],
        toolkits: [
          mcpToolkit(() => ({
            type: "streamable-http",
            url: stub.url,
          })),
        ],
      });

      const session = await agent.createSession({});

      try {
        const firstResponse = await session.run({
          input: [
            {
              type: "message",
              role: "user",
              content: [{ type: "text", text: "What's running tonight?" }],
            },
          ],
        });

        const expectedFirst: AgentResponse = {
          content: [{ type: "text", text: "Ready to roll." }],
          output: [
            {
              type: "model",
              content: [
                {
                  type: "tool-call",
                  tool_name: "list_shuttles",
                  tool_call_id: "call_1",
                  args: { shift: "evening" },
                },
              ],
            },
            {
              type: "tool",
              tool_name: "list_shuttles",
              tool_call_id: "call_1",
              input: { shift: "evening" },
              output: [
                { type: "text", text: "Shuttle summary for evening shift." },
                {
                  type: "image",
                  mime_type: "image/png",
                  data: IMAGE_DATA,
                },
                {
                  type: "audio",
                  data: AUDIO_DATA,
                  format: "mp3",
                },
              ],
              is_error: false,
            },
            {
              type: "model",
              content: [{ type: "text", text: "Ready to roll." }],
            },
          ],
        };
        t.assert.deepStrictEqual(firstResponse, expectedFirst);

        stub.updateTool({
          name: "list_shuttles_v2",
          description: "List active shuttle routes with live updates",
          responder: ({ shift }) => ({
            content: [
              {
                type: "text",
                text: `Updated shuttle roster for ${shift} shift.`,
              },
            ],
          }),
        });

        await new Promise((resolve) => setTimeout(resolve, 20));

        const secondResponse = await session.run({
          input: [
            {
              type: "message",
              role: "user",
              content: [{ type: "text", text: "How about now?" }],
            },
          ],
        });

        const expectedSecond: AgentResponse = {
          content: [{ type: "text", text: "Routes synced." }],
          output: [
            {
              type: "model",
              content: [
                {
                  type: "tool-call",
                  tool_name: "list_shuttles_v2",
                  tool_call_id: "call_2",
                  args: { shift: "evening" },
                },
              ],
            },
            {
              type: "tool",
              tool_name: "list_shuttles_v2",
              tool_call_id: "call_2",
              input: { shift: "evening" },
              output: [
                {
                  type: "text",
                  text: "Updated shuttle roster for evening shift.",
                },
              ],
              is_error: false,
            },
            {
              type: "model",
              content: [{ type: "text", text: "Routes synced." }],
            },
          ],
        };
        t.assert.deepStrictEqual(secondResponse, expectedSecond);
      } finally {
        await session.close();
      }
    } finally {
      await stub.stop();
    }
  });

  test("convertMCPContentToParts maps supported content", (t: TestContext) => {
    const contents: MCPContent[] = [
      { type: "text", text: "sample" },
      { type: "image", mimeType: "image/png", data: IMAGE_DATA },
      { type: "audio", mimeType: "audio/mpeg", data: AUDIO_DATA },
      {
        // @ts-expect-error Testing that unsupported types are skipped.
        type: "resource_link",
        uri: "https://example.com",
        name: "ignored",
      },
    ];

    const parts = convertMCPContentToParts(contents);

    t.assert.deepStrictEqual(parts, [
      { type: "text", text: "sample" },
      {
        type: "image",
        mime_type: "image/png",
        data: IMAGE_DATA,
      },
      {
        type: "audio",
        data: AUDIO_DATA,
        format: "mp3",
      },
    ]);
  });
});
